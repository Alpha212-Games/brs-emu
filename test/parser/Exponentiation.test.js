const Parser = require("../../lib/parser");
const Expr = require("../../lib/parser/Expression");
const { Lexeme } = require("../../lib/Lexeme");
const OrbsError = require("../../lib/Error");

const { token, EOF } = require("./ParserTests");

describe("parser", () => {
    afterEach(() => OrbsError.reset());

    describe("exponentiation expressions", () => {
        it("parses exponentiation operators", () => {
            let parsed = Parser.parse([
                token(Lexeme.Integer, 2),
                token(Lexeme.Caret),
                token(Lexeme.Integer, 3),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses repeated exponentiation as left-associative", () => {
            let parsed = Parser.parse([
                token(Lexeme.Integer, 2),
                token(Lexeme.Caret),
                token(Lexeme.Integer, 3),
                token(Lexeme.Caret),
                token(Lexeme.Integer, 4),
                EOF
            ]);
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    });
});