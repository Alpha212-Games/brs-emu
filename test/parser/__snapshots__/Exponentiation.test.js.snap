// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser exponentiation expressions parses exponentiation operators 1`] = `
Binary {
  "left": Literal {
    "value": 2,
  },
  "right": Literal {
    "value": 3,
  },
  "token": Object {
    "kind": 6,
    "line": 1,
    "literal": undefined,
  },
}
`;

exports[`parser exponentiation expressions parses repeated exponentiation as left-associative 1`] = `
Binary {
  "left": Binary {
    "left": Literal {
      "value": 2,
    },
    "right": Literal {
      "value": 3,
    },
    "token": Object {
      "kind": 6,
      "line": 1,
      "literal": undefined,
    },
  },
  "right": Literal {
    "value": 4,
  },
  "token": Object {
    "kind": 6,
    "line": 1,
    "literal": undefined,
  },
}
`;
